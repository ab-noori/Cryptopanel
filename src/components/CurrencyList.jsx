import React from 'react';
import PropTypes from 'prop-types';
import { useNavigate } from 'react-router-dom';
import millify from 'millify';
import './CurrencyList.scss';

const CurrencyList = ({
  currencies, query, handleQuery, handleKeyDown,
}) => {
  const navigate = useNavigate();

  const handleRowClick = (currencyId) => {
    navigate(`/Cryptopanel/${currencyId}`);
  };

  return (
    <>
      <input
        type="text"
        value={query}
        placeholder="Search"
        onChange={handleQuery}
        onKeyDown={handleKeyDown}
      />

      <div className="table-container">
        <table>
          <thead>
            <tr>
              <th>Rank</th>
              <th>Name</th>
              <th>Symbol</th>
              <th>Current Price</th>
              <th>24h Change</th>
              <th>Total Volume</th>
              <th>Market Cap</th>
            </tr>
          </thead>
          <tbody>
            {currencies.map((currency) => (
              <tr key={currency.id} onClick={() => handleRowClick(currency.id)}>
                <td data-label="Rank">{millify(currency?.market_cap_rank)}</td>
                <td data-label="Name">
                  <img
                    src={currency?.image}
                    alt={`${currency?.image}-logo`}
                    className="img"
                  />
                  {currency?.name}
                </td>
                <td data-label="Symbol">{currency?.symbol}</td>
                <td data-label="Current Price">{`$ ${millify(currency?.current_price)}`}</td>
                <td data-label="24h Change">{`${millify(currency?.price_change_percentage_24h)} %`}</td>
                <td data-label="Total Volume">{`$ ${millify(currency?.total_volume)}`}</td>
                <td data-label="Market Cap">{`$ ${millify(currency?.market_cap)}`}</td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    </>
  );
};

CurrencyList.propTypes = {
  currencies: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      rank: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      image: PropTypes.string.isRequired,
      symbol: PropTypes.string.isRequired,
      currentPrice: PropTypes.number.isRequired,
      oneHourPriceChange: PropTypes.number.isRequired,
      oneDayPriceChange: PropTypes.number.isRequired,
      oneWeekPriceChange: PropTypes.number.isRequired,
      totalVolume: PropTypes.number.isRequired,
      marketCap: PropTypes.number.isRequired,
    }),
  ).isRequired,
  query: PropTypes.string.isRequired,
  handleQuery: PropTypes.func.isRequired,
  handleKeyDown: PropTypes.func.isRequired,
};

export default CurrencyList;
