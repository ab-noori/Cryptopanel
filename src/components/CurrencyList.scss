table {
  width: 98%;
  margin: 1%;
  border-collapse: collapse;
}

th,
td {
  padding: 0.5rem;
  text-align: center;
  border-bottom: 1px solid #ddd;
}

/* Added CSS for striped effect on table headers */
thead th {
  background-color: #f2f2f2;
}

td {
  cursor: pointer;
}

th:nth-child(1),
td:nth-child(1),
th:nth-child(2),
td:nth-child(2) {
  text-align: left;
}

td img {
  width: 20px;
  margin: 0 10px 0 0;
}

tr:nth-child(even) {
  background-color: #f2f2f2;
}

tr:hover {
  background-color: #eaeaea;
}

thead th:first-child {
  border-bottom: none;
}

/* Rest of the CSS for responsive layout */
@media screen and (max-width: 768px) {
  /* Adjust the table layout for smaller screens */
  table {
    font-size: 12px;
    display: block;
    overflow-x: auto;
    padding: 5px;
  }

  thead {
    display: none;
  }

  tbody {
    display: flex;
    justify-content: space-between;
    flex-wrap: wrap;
  }

  /* Remove Rank column in mobile view */
  tr {
    display: flex;
    flex-direction: column;
    width: 50%;
    justify-content: space-between;
  }

  /* Increase image size */
  td img {
    width: 50px;
    margin: 0;
  }

  th {
    display: none;
  }

  td {
    position: relative;
    padding-left: 60%;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    border-bottom: 1px solid #ddd;
    padding-bottom: 10px;
  }

  td::before {
    position: absolute;
    top: 0.5rem;
    left: 0.5rem;
    content: attr(data-label);
    font-weight: bold;
  }

  /* Alternate row colors */
  tbody tr:nth-child(4n),
  tbody tr:nth-child(4n + 1) {
    background-color: #e0dfdf;
  }

  tbody tr:nth-child(4n + 2),
  tbody tr:nth-child(4n + 3) {
    background-color: #ececec;
  }

  th:nth-child(1),
  td:nth-child(1) {
    display: none;
  }

  th:nth-child(2),
  td:nth-child(2) {
    display: flex;
    justify-content: center;
    padding: 10px 0;
    font-size: 0;
  }

  /* Hover effect for mobile view */
  tbody tr:hover {
    background-color: #c9c9c9;
  }
}
