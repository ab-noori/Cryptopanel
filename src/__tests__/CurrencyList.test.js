import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import CurrencyList from '../components/CurrencyList';

test('renders CurrencyList component correctly', () => {
  const currencies = [
    {
      id: 1,
      rank: 1,
      name: 'Bitcoin',
      image: 'bitcoin.png',
      symbol: 'BTC',
      currentPrice: 50000,
      oneHourPriceChange: 0.5,
      oneDayPriceChange: 1.2,
      oneWeekPriceChange: -0.8,
      totalVolume: 1000000,
      marketCap: 1000000000,
    },
    // Add more sample data as needed
  ];

  const { container } = render(
    <Router>
      <CurrencyList
        currencies={currencies}
        query=""
        handleQuery={() => {}}
        handleKeyDown={() => {}}
      />
    </Router>,
  );

  expect(container).toMatchSnapshot();
});
