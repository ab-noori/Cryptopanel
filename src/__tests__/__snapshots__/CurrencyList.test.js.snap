// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders CurrencyList component correctly 1`] = `
<div>
  <div
    class="search-bar"
  >
    <input
      placeholder="Search"
      type="text"
      value=""
    />
    <button
      type="button"
    >
      Search
    </button>
  </div>
  <div
    class="table-container"
  >
    <table>
      <thead>
        <tr>
          <th>
            Rank
          </th>
          <th>
            Name
          </th>
          <th>
            Symbol
          </th>
          <th>
            Current Price
          </th>
          <th>
            24h Change
          </th>
          <th>
            Total Volume
          </th>
          <th>
            Market Cap
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            data-label="Rank"
          >
            undefined
          </td>
          <td
            data-label="Name"
          >
            <img
              alt="bitcoin.png-logo"
              class="img"
              src="bitcoin.png"
            />
            Bitcoin
          </td>
          <td
            data-label="Symbol"
          >
            BTC
          </td>
          <td
            data-label="Current Price"
          >
            $ undefined
          </td>
          <td
            data-label="24h Change"
          >
            undefined %
          </td>
          <td
            data-label="Total Volume"
          >
            $ undefined
          </td>
          <td
            data-label="Market Cap"
          >
            $ undefined
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
